File: 백준/Gold/14499. 주사위 굴리기/주사위 굴리기.py
import sys
input = sys.stdin.readline

n, m, x, y, k = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]
order = list(map(int, input().split()))

dice = [[0]*3 for _ in range(4)]
dx = [0, 0, 0, -1, 1]
dy = [0, 1, -1, 0, 0]

for i in order:
    nx, ny = x + dx[i], y + dy[i]
    if nx < 0 or nx >= n or ny < 0 or ny >= m:
        continue
     
    # 주사위 굴리기
    if i == 1: # 동쪽
        dice[0][1], dice[1][0], dice[1][2], dice[2][1] = dice[1][2], dice[0][1], dice[2][1], dice[1][0]
    elif i == 2: # 서쪽
        dice[0][1], dice[1][0], dice[1][2], dice[2][1] = dice[1][0], dice[2][1], dice[0][1], dice[1][2]
    elif i == 3: #북쪽
        dice[0][1], dice[1][1], dice[2][1], dice[3][1] = dice[3][1], dice[0][1], dice[1][1], dice[2][1]
    else:
        dice[0][1], dice[1][1], dice[2][1], dice[3][1] = dice[1][1], dice[2][1], dice[3][1], dice[0][1]
    
    if graph[nx][ny] == 0:
        graph[nx][ny] = dice[2][1] # 바닥면
    else:
        dice[2][1] = graph[nx][ny]
        graph[nx][ny] = 0

    print(dice[0][1]) # 윗 면
    x, y = nx, ny

File: 백준/Gold/14499. 주사위 굴리기/주사위 굴리기.js
const input = require("fs")
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt"
  )
  .toString()
  .split("\n");

let index = 0;
const [n, m, xx, yy, k] = input[index++].split(" ").map(Number);

const board = [];
for (let i = 0; i < n; i++) {
  board.push(input[index++].split(" ").map(Number));
}

const commands = input[index++].split(" ").map(Number);
let dice = [
  [0, 0, 0],
  [0, 0, 0],
  [0, 0, 0],
  [0, 0, 0],
];

let x = xx;
let y = yy;

const dx = [0, 0, 0, -1, 1];
const dy = [0, 1, -1, 0, 0];

const moves = [0, move_right, move_left, move_up, move_down];

const answer = [];

for (let i = 0; i < k; i++) {
  const cur_d = commands[i];
  const nx = x + dx[cur_d];
  const ny = y + dy[cur_d];

  if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;

  const moved_dice = moves[cur_d](dice);

  if (board[nx][ny] === 0) {
    board[nx][ny] = moved_dice[3][1];
  } else {
    moved_dice[3][1] = board[nx][ny];
    board[nx][ny] = 0;
  }

  answer.push(moved_dice[1][1]);
  dice = moved_dice.map((d) => d.slice());
  x = nx;
  y = ny;
}

console.log(answer.join("\n"));

function move_right(dice) {
  const moved_dice = dice.map((d) => d.slice());

  moved_dice[1][0] = dice[3][1];
  moved_dice[1][1] = dice[1][0];
  moved_dice[1][2] = dice[1][1];
  moved_dice[3][1] = dice[1][2];

  return moved_dice;
}

function move_left(dice) {
  const moved_dice = dice.map((d) => d.slice());

  moved_dice[1][0] = dice[1][1];
  moved_dice[1][1] = dice[1][2];
  moved_dice[1][2] = dice[3][1];
  moved_dice[3][1] = dice[1][0];

  return moved_dice;
}

function move_up(dice) {
  const moved_dice = dice.map((d) => d.slice());

  moved_dice[0][1] = dice[1][1];
  moved_dice[1][1] = dice[2][1];
  moved_dice[2][1] = dice[3][1];
  moved_dice[3][1] = dice[0][1];

  return moved_dice;
}

function move_down(dice) {
  const moved_dice = dice.map((d) => d.slice());

  moved_dice[0][1] = dice[3][1];
  moved_dice[1][1] = dice[0][1];
  moved_dice[2][1] = dice[1][1];
  moved_dice[3][1] = dice[2][1];

  return moved_dice;
}
